### PR TITLE
feat: add new perk Hidden Pockets

### PR DESCRIPTION
**Pull request by Darxo**
### Screenshot

image

image

image

### CODE REVIEW

File: mod_reforged/config/z_nested_tooltips.nut [hunk 1]
```diff
@@ -21,6 +21,7 @@
...
 			Round = ::MSU.Class.BasicTooltip("Round", ::Reforged.Mod.Tooltips.parseString("Combat in battle brothers is turn-based. Each combat consists of a series of rounds. During a round, characters act in [turns|Concept.Turn]. A round ends when all characters have ended their [turns|Concept.Turn].")),
 			ZoneOfControl = ::MSU.Class.BasicTooltip("Zone of Control", ::Reforged.Mod.Tooltips.parseString("Most melee characters exert Zone of Control on their surrounding tiles. Trying to move out of enemy Zone of Control will trigger one free attack from each controlling enemy until the first attack hit or every attack missed. A hit will cancel the movement.")),
+			BagSlots = ::MSU.Class.BasicTooltip("Bag Slots", ::Reforged.Mod.Tooltips.parseString("Bagslots can be used to store additional weapons or utility items for use during combat. Every character has 2 Bagslots by default and can have at most 4 Bagslots.")),
```
Comment (LordMidas) [inline mod_reforged/config/z_nested_tooltips.nut:24]: - `Bag slots` with a space.
- The last sentence can end as `and can have a maximum of 4.`
- Both the default and the max number of bag slots should be dynamically read from `item_container.m.UnlockedBagSlots` and `::Const.ItemSlotSpaces` respectively. For the value from `item_container` you can instantiate an `item_container` locally above in this file and store the result of `.m.UnlockedBagSlots` in a local variable.

File: mod_reforged/hooks/config/strings.nut [hunk 2]
```diff
@@ -118,6 +118,7 @@
 	RF_Ghostlike = "Ghostlike",
 	RF_HaleAndHearty = "Hale and Hearty",
 	RF_Heft = "Heft",
+	RF_HiddenPockets = "Hidden Pockets"
```
Comment (LordMidas) [inline mod_reforged/hooks/config/strings.nut:121]: Missing comma

File: mod_reforged/hooks/config/strings.nut [hunk 3]
```diff
@@ -1257,6 +1270,20 @@ foreach (vanillaDesc in vanillaDescriptions)
 			]
 		}]
  	}),
+	RF_HiddenPockets = ::UPD.getDescription({
+		Effects = [{
```
Comment (LordMidas) [inline mod_reforged/hooks/config/strings.nut:1274]: I'd improve the formatting here by pushing the `{` to the next line and similarly at the end pushing the `]` to the next line. Then the tables can be indented one tab to the right.

File: mod_reforged/hooks/skills/actives/throw_dirt_skill.nut [hunk 4]
```diff
@@ -0,0 +1,22 @@
+::Reforged.HooksMod.hook("scripts/skills/actives/throw_dirt_skill", function(q) {
+	q.create = @(__original) function()
+	{
+		__original();
+		this.m.Description = "This character hides a substantial amount of sand in their pockets, ready to be unleashed on an attacker.";
```
Comment (LordMidas) [inline mod_reforged/hooks/skills/actives/throw_dirt_skill.nut:5]: `unleashed` isn't appropriate here for throwing sand. Similarly `attacker`. Could say `ready to be thrown at someone.`

File: mod_reforged/hooks/skills/actives/throw_dirt_skill.nut [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+	// Vanilla does not provide an implementation for this function. Since we want to display the cost of the skill we can't build upon the base function from skill.nut
+	q.getTooltip = @() function()
+	{
+		local ret = this.skill.getDefaultUtilityTooltip();
+
+		ret.push({
+			id = 5,
+			type = "text",
+			icon = "ui/icons/special.png",
+			text = ::Reforged.Mod.Tooltips.parseString("Applies the [Distracted|Skill+distracted_effect] effect to the target")
```
Comment (LordMidas) [inline mod_reforged/hooks/skills/actives/throw_dirt_skill.nut:17]: `on the target`

File: mod_reforged/hooks/skills/actives/throw_dirt_skill.nut [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+		})
```
Comment (LordMidas) [inline mod_reforged/hooks/skills/actives/throw_dirt_skill.nut:18]: Missing semi-colon.

File: scripts/skills/actives/rf_pocket_sand_skill.nut [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+				text = "Remaining uses: " + ::MSU.Text.colorGreen(this.m.RemainingUses)
+			})
+		}
+
+		return ret;
+	}
+
+	function isUsable()
+	{
+		return (this.throw_dirt_skill.isUsable() && this.m.RemainingUses != 0);
```
Comment (LordMidas) [inline scripts/skills/actives/rf_pocket_sand_skill.nut:52]: No need for parantheses with a return statement like this.

File: scripts/skills/actives/rf_pocket_sand_skill.nut [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+// New Functions
+	function getRemainingUses()
+	{
+		if (this.getContainer().getActor().isPlacedOnMap())
+		{
+			return this.m.RemainingUses;
+		}
+		else	// This way the skill will give better feedback while outside of combat
```
Comment (LordMidas) [inline scripts/skills/actives/rf_pocket_sand_skill.nut:76]: But you don't actually use this function outside of combat. The logic in these two functions and `getTooltip` needs to be updated so that outside of combat it always shows the maximum number of uses.

File: scripts/skills/perks/perk_rf_hidden_pockets.nut [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+		this.m.Type = ::Const.SkillType.Perk;
+		this.m.Order = ::Const.SkillOrder.Last;
+		this.m.IsActive = false;
+		this.m.IsStacking = false;
+		this.m.IsHidden = false;
+	}
+
+	function onUpdate( _properties )
+	{
+		// This works exactly like how the vanilla bags and belts perk adds its slots. Both at the same time will therefor not have any effect
```
Comment (LordMidas) [inline scripts/skills/perks/perk_rf_hidden_pockets.nut:18]: `therefore`

File: scripts/skills/perks/perk_rf_hidden_pockets.nut [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+			items.removeFromBag(item);
+			this.World.Assets.getStash().add(item);
+		}
+
+		item = items.getItemAtBagSlot(3);
+
+		if (item != null)
+		{
+			items.removeFromBag(item);
+			this.World.Assets.getStash().add(item);
```
Comment (LordMidas) [inline scripts/skills/perks/perk_rf_hidden_pockets.nut:42]: Replace `this.World` with `::World` even though it's a copy of the vanilla function.

File: scripts/skills/perks/perk_rf_hidden_pockets.nut [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+		}
+
+		// This line exists in bags and belts aswell. Let's hope that there happens another onUpdate over all skills right after.
```
Comment (LordMidas) [inline scripts/skills/perks/perk_rf_hidden_pockets.nut:45]: `as well`

File: scripts/skills/perks/perk_rf_hidden_pockets.nut [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+		// Otherwise this can cause problems with both Bags and Belts and Hidden Pockets were allocated, but only one was removed
+		this.getContainer().getActor().getItems().setUnlockedBagSlots(2);
+	}
+
+	function onAdded()
+	{
+		this.getContainer().add(::new("scripts/skills/actives/rf_pocket_sand_skill"));
+	}
```
Comment (LordMidas) [inline scripts/skills/perks/perk_rf_hidden_pockets.nut:53]: I'd move this function up above `onRemoved` because that function removes this skill and this one adds it so it makes more sense logically when viewing the perk's code.

File: scripts/skills/perks/perk_rf_hidden_pockets.nut [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+		item = items.getItemAtBagSlot(3);
+
+		if (item != null)
+		{
+			items.removeFromBag(item);
+			this.World.Assets.getStash().add(item);
+		}
+
+		// This line exists in bags and belts aswell. Let's hope that there happens another onUpdate over all skills right after.
+		// Otherwise this can cause problems with both Bags and Belts and Hidden Pockets were allocated, but only one was removed
```
Comment (LordMidas) [inline scripts/skills/perks/perk_rf_hidden_pockets.nut:46]: I'd suggest to please test this in-game and find out exactly what happens in this case. And if something unexpected happens then we can try to fix it.

File: scripts/skills/perks/perk_rf_hidden_pockets.nut [hunk 14]
```diff
@@ -0,0 +1,54 @@
+this.perk_rf_hidden_pockets <- ::inherit("scripts/skills/skill", {
+	m = {},
+	function create()
+	{
+		this.m.ID = "perk.rf_hidden_pockets";
+		this.m.Name = ::Const.Strings.PerkName.RF_HiddenPockets;
+		this.m.Description = ::Const.Strings.PerkDescription.RF_HiddenPockets;
+		this.m.Icon = "ui/perks/rf_hidden_pockets.png";
+		this.m.Type = ::Const.SkillType.Perk;
+		this.m.Order = ::Const.SkillOrder.Last;
```
Comment (LordMidas) [inline scripts/skills/perks/perk_rf_hidden_pockets.nut:10]: Why Last?

File: scripts/skills/actives/rf_pocket_sand_skill.nut [hunk 15]
```diff
@@ -0,0 +1,87 @@
+this.rf_pocket_sand_skill <- this.inherit("scripts/skills/actives/throw_dirt_skill", {
+	m = {
+		RemainingUses = 0
+	},
+
```
Comment (LordMidas) [inline scripts/skills/actives/rf_pocket_sand_skill.nut:5]: Vanilla doesn't put an empty space between the `m` table and `function create` so let's keep it like that.

File: mod_reforged/hooks/config/strings.nut [hunk 16]
```diff
@@ -289,7 +290,19 @@ local vanillaDescriptions = [
...
 	},
-
+	{
+		ID = "perk.bags_and_belts",
+		Key = "BagsAndBelts",
+		Description = ::UPD.getDescription({
+	 		Effects = [{
+ 				Type = ::UPD.EffectType.Passive,
+ 				Description = [
+					"Unlock two extra [Bag Slots|Concept.BagSlots].",
```
Comment (LordMidas) [inline mod_reforged/hooks/config/strings.nut:300]: `bag slots` both words can be decapitalized.

File: mod_reforged/hooks/config/strings.nut [hunk 17]
```diff
@@ -1257,6 +1270,20 @@ foreach (vanillaDesc in vanillaDescriptions)
 			]
 		}]
  	}),
+	RF_HiddenPockets = ::UPD.getDescription({
+		Effects = [{
+			Type = ::UPD.EffectType.Passive,
+			Description = [
+				"Unlock two extra [Bag Slots|Concept.BagSlots]."
```
Comment (LordMidas) [inline mod_reforged/hooks/config/strings.nut:1277]: `bag slots` both words can be decapitalized.

File: scripts/skills/actives/rf_pocket_sand_skill.nut [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+	function create()
+	{
+		this.throw_dirt_skill.create();
+
+		this.m.ID = "actives.rf_pocket_sand_skill";
+		this.m.Name = "Throw Pocket Sand";
```
Comment (LordMidas) [inline scripts/skills/actives/rf_pocket_sand_skill.nut:11]: I'd say, in a separate commit, also add this skill to the `PossibleSkills` of the same ai behavior that uses throw_dirt. And then add that here as `AIBehaviorID`. This way this skill can be given to enemies as well.

**Review by LordMidas — CHANGES_REQUESTED** [review]

File: scripts/skills/perks/perk_rf_hidden_pockets.nut (hunk 13 above)
Comment (Darxo) [reply scripts/skills/perks/perk_rf_hidden_pockets.nut]: There is no way in vanilla or reforged to unlearn a single perk point so this can never occur. Even if there was, this would still be a rare edge case. 

I don't plan on testing for this rare edge case. If you feel better I can remove this comment

Comment (LordMidas) [reply scripts/skills/perks/perk_rf_hidden_pockets.nut]: No, please test it. It isn't an elaborate test:
- Add both perks via Taro's dev console.
- Fill the bags with items. 
- Then remove bags and belts. 
- Repeat the test but this time remove Trickster's Purses instead.

File: mod_reforged/hooks/skills/actives/throw_dirt_skill.nut [hunk 19]
```diff
@@ -0,0 +1,36 @@
+::Reforged.HooksMod.hook("scripts/skills/actives/throw_dirt_skill", function(q) {
+	q.create = @(__original) function()
+	{
+		__original();
+		this.m.Description = "This character hides a substantial amount of sand in their pockets, ready to be ready to be thrown at someone.";
```
Comment (LordMidas) [inline mod_reforged/hooks/skills/actives/throw_dirt_skill.nut:5]: `ready to be` is repeated

File: mod_reforged/hooks/skills/actives/throw_dirt_skill.nut [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+	}
+
+	q.onVerifyTarget = @(__original) function( _originTile, _targetTile )
+	{
+		local ret = __original(_originTile, _targetTile);
+
+		if (ret && _targetTile.getEntity().getSkills().hasSkill("effects.distracted"))	// Throw Dirt can no longer be used one someone who is already affected by it
```
Comment (LordMidas) [inline mod_reforged/hooks/skills/actives/throw_dirt_skill.nut:12]: `on someone` typo

File: mod_reforged/hooks/skills/actives/throw_dirt_skill.nut [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+		{
+			return false;
+		}
+		else
+		{
+			return ret;
+		}
```
Comment (LordMidas) [inline mod_reforged/hooks/skills/actives/throw_dirt_skill.nut:19]: This is needlessly complicated to follow through logically with all the if/else. This can simply be the following. Please use this:
```squirrel
return __original(_originTile, _targetTile) && !_targetTile.getEntity().getSkills().hasSkill("effects.distracted");
```

File: scripts/skills/actives/rf_pocket_sand_skill.nut [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+				});
+			}
+		}
+		else
+		{
+			ret.push({
+				id = 21,
+				type = "text",
+				icon = "ui/icons/special.png",
+				text = "You have one use for each empty bag slot that you start the battle with",
```
Comment (LordMidas) [inline scripts/skills/actives/rf_pocket_sand_skill.nut:45]: `bag slot` should be hyperlinked

File: scripts/skills/actives/rf_pocket_sand_skill.nut [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+			});
+		}
+
+		return ret;
+	}
+
+	function isUsable()
+	{
+		return this.throw_dirt_skill.isUsable() && (this.m.RemainingUses != 0);
```
Comment (LordMidas) [inline scripts/skills/actives/rf_pocket_sand_skill.nut:54]: Please remove the redundant parantheses

File: scripts/skills/actives/rf_pocket_sand_skill.nut [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+		this.throw_dirt_skill.onCombatStarted();
+
+		this.m.RemainingUses = this.getMaximumUses();
+	}
+
+	function onUse( _user, _targetTile )
+	{
+		this.throw_dirt_skill.onUse(_user, _targetTile);
+
+		this.m.RemainingUses = ::Math.max(0, this.m.RemainingUses - 1);
```
Comment (LordMidas) [inline scripts/skills/actives/rf_pocket_sand_skill.nut:68]: return the value returned by the call to the parent's `onUse`

File: mod_reforged/hooks/config/strings.nut [hunk 25]
```diff
@@ -1382,6 +1395,22 @@ foreach (vanillaDesc in vanillaDescriptions)
...
+			{
+				Type = ::UPD.EffectType.Passive,
+				Description = [
+					"Unlock two extra [bag slots|Concept.BagSlots].",
+				],
+			},
+			{
+				Type = ::UPD.EffectType.Active,
+				Description = [
+					"Unlocks the [Pocket Sand|Skill+rf_pocket_sand_skill] skill which allows you to store sand in your empty bag slots to distract nearby enemies with.",
```
Comment (LordMidas) [inline mod_reforged/hooks/config/strings.nut:1409]: `bag slots` should be hyperlinked.

File: mod_reforged/hooks/config/strings.nut [hunk 26]
```diff
@@ -118,6 +118,7 @@
 	RF_Ghostlike = "Ghostlike",
 	RF_HaleAndHearty = "Hale and Hearty",
 	RF_Heft = "Heft",
+	RF_TrickstersPurses = "Tricksters Purses",
```
Comment (LordMidas) [inline mod_reforged/hooks/config/strings.nut:121]: `Trickster\'s`

File: mod_reforged/hooks/config/strings.nut (hunk 16 above)
Comment (LordMidas) [reply mod_reforged/hooks/config/strings.nut:325]: Unresolving this as this was missed.

**Review by LordMidas — CHANGES_REQUESTED** [review]

File: mod_reforged/hooks/config/strings.nut (hunk 16 above)
Comment (LordMidas) [reply mod_reforged/hooks/config/strings.nut:325]: Unresolving this as this was missed in the changes.

File: scripts/skills/actives/rf_pocket_sand_skill.nut (hunk 23 above)
Comment (Darxo) [reply scripts/skills/actives/rf_pocket_sand_skill.nut:54]: I use these redundant parantheses to improve readability. 
Please provide me with a hard rule about this so I know when I'm allowed to put optional parantheses

File: scripts/skills/perks/perk_rf_hidden_pockets.nut (hunk 13 above)
Comment (Darxo) [reply scripts/skills/perks/perk_rf_hidden_pockets.nut]: I tested that out now.
The result is actually not suprising:
- If I'm in the character/inventory screen and refund a random perk (e.g. with that suggested potion) then nothing happens at first.
- Once I reopen the menu **or** equip/unequip any item to the brother who just got backs&pelts or tricksters purses unlearned then suddenly he unequips both items in the slot 3 and 4 of his bags. Even if he still has 4 slots afterwards. This is unexpected behavior from the `onRemoved` function of both perks, they don't check if the respective other perk is present

So as of right now this does not void any items, unless you unlearn such a perk with a full inventory, but that already happens in vanilla when using the oblivion potion. The observed behavior is a bit weird but does not break anything at the moment.

### Future

We should look into managing the unlocked bagslots in a more central location, currently that is done within the `item_container`.

Maybe we can use `bag_fatigue` skill to manage the unlocked bagslots.
Or we move the unlocked bagslots into the `character.nut` which means that it can now be manipulated during the update by changing the value in `_properties`

Comment (LordMidas) [reply scripts/skills/perks/perk_rf_hidden_pockets.nut]: Thank you for testing. Yes it is sufficient to update the comment so we know what to expect. We don't have to implement any fix for this behavior.

**Review by LordMidas — CHANGES_REQUESTED** [review]

File: scripts/skills/perks/perk_rf_tricksters_purses.nut [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+
+	function onAdded()
+	{
+		this.getContainer().add(::new("scripts/skills/actives/rf_pocket_sand_skill"));
+	}
+
+	function onRemoved()
+	{
+		if (this.getContainer().getAcotr().getItems().getUnlockedBagSlots())
+		this.getContainer().removeByID("actives.rf_pocket_sand_skill");
```
[the remark this answers is not in the document]
Comment (Darxo) [reply scripts/skills/perks/perk_rf_tricksters_purses.nut:26]: Must be a leftover my me trying to fix the presents of both perks and removal of one. But that is for another time so I will remove the leftover.

**Review by LordMidas — APPROVED** [review]
